cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlManager.java

```diff
@@ -0,0 +1,244 @@
+import org.eclipse.jgit.lib.Ref;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A Source Control Manager that is responsible for handling interfacing with git. It fetches credentials
+ * from a {@link io.cdap.cdap.api.security.store.SecureStore}, builds an {@link AuthStrategy} for different
+ * Git hosting services and provides version control operations.
+ */
+public class SourceControlManager implements AutoCloseable {
+  private static final Logger LOG = LoggerFactory.getLogger(SourceControlManager.class);
+  private final SourceControlContext context;
+  private final AuthStrategy authStrategy;
+  private final File localRepoDirectory;
+  private final File validationRepoDirectory;
+  private final int gitCommandTimeoutSeconds;
+  private Git git;
+
+  public SourceControlManager(SourceControlContext context) throws AuthStrategyNotFoundException {
+    this.context = context;
+    this.authStrategy = getAuthStrategy();
+    this.localRepoDirectory = context.getRepoBasePath().toFile();
+    this.gitCommandTimeoutSeconds = context.getGitCommandTimeoutSeconds();
+    this.validationRepoDirectory = new File("/tmp/" + UUID.randomUUID());
+  }
+
+  // Deletes a directory by recursively deleting its contents.
+  private boolean deleteDirectory(File dir) {
+    File[] allContents = dir.listFiles();
+    if (allContents != null) {
+      for (File file : allContents) {
+        deleteDirectory(file);
+      }
+    }
+    return dir.delete();
+  }
+
+  private <C extends TransportCommand> C getAuthenticatedCommandWithTimeout(C command) throws AuthenticationException {
+    command.setCredentialsProvider(authStrategy.getCredentialProvider(context));
+    Optional<TransportConfigCallback> transport = authStrategy.getTransportConfigCallback(context);
+    transport.ifPresent(command::setTransportConfigCallback);
+    command.setTimeout(gitCommandTimeoutSeconds);
+    return command;
+  }
+
+  /**
+   * Initialises the Git repository by cloning remote.
+   *
+   * @throws GitAPIException         when a Git operation fails.
+   * @throws AuthenticationException when correct authentication credentials are not available.
+   * @throws IOException             when file or network I/O fails.
+   */
+  public void initialise() throws GitAPIException, AuthenticationException, IOException {
+    if (git != null) {
+      return;
+    }
+    // Clean up the directory if it already exists.
+    if (localRepoDirectory.exists() && !deleteDirectory(localRepoDirectory)) {
+      throw new IOException("Failed to clean up local repository directory " + context.getRepoBasePath());
+    }
+    String branch = "refs/heads/" + context.getRepositoryConfig().getDefaultBranch();
```

Comment:
Since in the discussion, `defaultBranch` in config is optional, we might need to fetch the default branch from provider (here it's GitHub) when `getDefaultBranch()` returns null

Reply:
Done

Reply:
If we don't set the branch, the HEAD of the remote repository is checked out automatically. I'll be adding tests for this brahaviour when I add the push/pull methods.